### ISSUE TITLE
Add elliptical spatial sky models

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1663]
Currently we have radially symmetric sky models, such as e.g. [SkyDisk](http://docs.gammapy.org/dev/api/gammapy.image.models.SkyDisk.html) or [SkyGaussian](http://docs.gammapy.org/dev/api/gammapy.image.models.SkyGaussian.html)

We should add support to fit equivalent elliptical sky models, i.e. elongated variants of those models that are stretched in one direction. This is sometimes used, especially for the elliptical disk and Gaussian
One question is how to parameterise the ellipticity or elongation. There is the simple parametrisation with major and minor axis and a position angle. That is unstable, because the position angle is degenerate for near-round sources, and flips from zero to 360 deg. There are better, much more stable parametrisations as explained in Section 10.1 in the SExtractor manual [here](https://www.astromatic.net/pubsvn/software/sextractor/trunk/doc/sextractor.pdf) or also some Python code to measure ellipses in photutils [here](https://photutils.readthedocs.io/en/stable/isophote.html). There is a large body of literature and software packages for measuring ellipticity shear parameters from the astronomy gravitational lensing community; I could look up references, but as mentioned below I don't think we will go that direction in Gammapy anyways. I think the parametrisation they use is as given in cell 3 in the notebook [here](https://github.com/Asterics2020-Obelics/School2018/blob/master/machinelearning/Data.ipynb).

Another question is whether to introduce separate model classes for elliptical models, or modify the existing ones in Gammapy to allow elongation (probably configured to be round by default, i.e. opt-in to fit elongation).

The final question is whether to work in the small angle approximation or fully on the sphere, and how to do the position-angle related computations exactly.

Here's what other have:
- Astropy has [Ellipse2D](http://docs.astropy.org/en/stable/api/astropy.modeling.functional_models.Ellipse2D.html) and [Gaussian2D](http://docs.astropy.org/en/stable/api/astropy.modeling.functional_models.Gaussian2D.html) with parametrisation major and minor axis and position angle. In a 2D plane, not on the sphere.
- Sherpa has [gauss2d](http://cxc.harvard.edu/sherpa/ahelp/gauss2d.html) with one extension, an ellipticity epsilon, and a position angle. In a 2D plane, not on the sphere.
- Gammalib has separate classes for the elliptical models:  [GModelSpatialEllipticalDisk](http://gammalib.sourceforge.net/doxygen/classGModelSpatialEllipticalDisk.html) and [GModelSpatialEllipticalGauss](http://gammalib.sourceforge.net/doxygen/classGModelSpatialEllipticalGauss.html). Those are defined on the sphere.
- It looks like astromodels also has an ellipse on the sphere, see [here](https://github.com/giacomov/astromodels/blob/1742826a9794ca7eef5f635b5448f4d289412389/astromodels/functions/functions_2D.py#L473).

My suggestion for Gammapy would be that we use the parametrisation semimajor / semiminor / position angle. About whether to add this to the existing spatial models or make new ones I'm not sure. If we amend the existing ones, they would almost become like the ones in Astropy, except that they would do some code to compute position angles on the sky instead of just x / y, because we evaluate on lon / lat. I think I have a preference for adding the elongated models as separate classes for now, and then only later to consider merging them with the existing ones. We might never want to do that for efficiency reasons, and if we make one class the evaluate would basically be an if / else that checks whether it's a round model or elongated, and dispatches to the evaluate calls for that.

So basically, the mission, should you choose to accept it, is to make a pull request for a first elliptical model of your choice, either:
- `gammapy.image.models.SkyEllipticalDisk`, which wraps the Astropy model (i.e. same parameters, and calls into [astropy.modeling.models.Ellipse2D.evaluate](http://docs.astropy.org/en/stable/_modules/astropy/modeling/functional_models.html#Ellipse2D.evaluate)
- `gammapy.image.models.SkyEllipticalGauss`, which wraps the Astropy model (i.e. same parameters, and calls into [astropy.modeling.models.Gaussian2D.evaluate](http://docs.astropy.org/en/stable/_modules/astropy/modeling/functional_models.html#Gaussian2D.evaluate)

**Comment by joleroi** [comment]
Note that this is not mission difficult, it's mission impossible. Difficult should be a walk in the park for you

**Comment by adonath** [comment]
This should be addressed in #1971.

**Comment by adonath** [comment]
@luca-giunti I think the elongated Gaussian is still missing, because #2024 has not been merged, but closed. OK to reopen?

**Comment by luca-giunti** [comment]
Yes! You're right, sorry

**Comment by luca-giunti** [comment]
Hi! As discussed at the Coding Sprint, I have implemented an elongated Gaussian model that is defined on the sphere, but, for the moment, normalized in small angle approximation.

Before opening yet another pull request on this topic, I have tested a bit what is the error we introduce by normalizing on the plane. In particular, in this [notebook](https://github.com/luca-giunti/share/blob/master/Untitled16.ipynb) I compare an elliptical Gaussian with null eccentricity with a radial Gaussian.  

If everyone agrees that the approximation is acceptable, I will open the PR. If other tests are needed, I can further play with the notebook.

Cheers!

**Comment by registerrier** [comment]
OK. Thanks @luca-giunti . This looks perfectly OK to go ahead then. A simple mention in the doc that the norm error is below 0.1% for radii lower than 5 deg is fine.

**Comment by adonath** [comment]
Thanks @luca-giunti! I agree the planar approximation for the norm is completely acceptable. 

Maybe spent ~1h to add a title and short abstract to the notebook and then add it to https://github.com/gammapy/gammapy-extra/tree/master/checks as a future reference?

**Comment by cdeil** [comment]
My suggestion was & is to do all the models in the planar approximation. That's e.g. what we do in astropy-regions or what DS9 does as well, and it's also consistent with how we do PSF convolution, and it's simpler and faster. But for sure, OK to do model eval partly in the planar approximation. Just make sure you put the formulae that you implement in the code in the docstring as well, and a sentence what it is. Will be nice to have elongated Gaussian, thanks!

**Comment by cdeil** [comment]
The elongated Gaussian was added by @luca-giunti in #2313 .

@adonath @luca-giunti - Let's wrap up this issue and polish the elongated sky models to v1.0 quality.

We could change nothing and just close. But there are a few things that I would change, to simplify and get easier to read and maintain code:
- Make ellipse and elongated Gaussian implementation more similar? I think ellipse could be rewritten to use a `sigma_eff` just like the Gaussian (a helper function called by both), the current code with the `offset_by` is a bit more complex and slower?
- Is it really useful to have separate classes for elongated / non-elongated? I think it was definitely nice for developing, to just add instead of edit existing classes, but now that it's figured out, I think merging the two Gaussian and the disk/ellipse could be a bit simpler? The default parameters would be non-elongated, and the implementation would have one `if`, to handle the round / elongated case, based on whether the elongation is zero or not. It's a matter of taste concerning one class with a bit more complexity, or two similar classes. In this case, especially from the user's perspective, I think a single class would be nicer. It's very typical to fit a round model, and then to free the elongation parameter and re-fit, and it's nice to be able to keep the same model object for the same source, especially if it's part of a large and complex linked model on several datasets.
- I would suggest to simplify the ellipse norm to just be the equivalent disk norm, just like what we have for Gaussian now. As you saw, the precision is very good even for huge sources as large as the field of view of CTA. This would allow us to get rid of the complex `compute_norm` with the trigonometry stuff and scipy.integrate.quad.

Thoughts?

**Comment by luca-giunti** [comment]
@cdeil I am applying the changes you suggest. I can easily unify the elongated and symmetric classes. However, I don't understand how to get rid of the scipy integration for the norm of the SkyEllipse.

You propose to have an ellipse normalization similar to the disk case, using `sigma_eff` instead of the disk radius (as in `norm = 1.0 / (2 * np.pi * (1 - np.cos(sigma_eff)))`), is that correct?

If so, we have a problem, because while this makes the model integrate to 1 (I checked), the normalization varies across the ellipse (since `sigma_eff` varies). 

So, I think that the `sigma_eff` can be introduced to get rid of the spherical trigonometry stuff and just define the "inside ellipse" condition as `in_ellipse = smooth_edge(sep - sigma_eff, edge)`. 
However, for the norm we still need to compute somehow the surface of the elliptical cap on the sphere - that's why I introduced the scipy integration.

Unless, you also propose to revert everything to planar approximation, in which case you easily have the area of the ellipse.

Am I missing something?

**Comment by cdeil** [comment]
I didn't go through the math, yes, planar approximation sounds good to me, it should be fine for normal use cases (sources up to a few deg in size max).

@adonath or @registerrier might disagree, so you could hold off on that change for now, to avoid changing back and forth.

**Comment by adonath** [comment]
After #2371, I think there is nothing left to do here. Closing now.
